# RNAseq Analysis - nfcore/rnaseq workflow

## nf-core/rnaseq pipeline - installation

<a href="https://nf-co.re/"><img src="./images/nf-core-logo.svg" width="300"></a>

__nf-core/rnaseq__ is a bioinformatics analysis pipeline used for RNA sequencing data. The pipeline is built using __Nextflow__, a workflow tool to run tasks across multiple compute infrastructures in a portable manner. Detailed documentation on how to install and run nextflow and nf-core bioinformatic pipelines can be found at the following sites:

- [https://nextflow.io/](https://nextflow.io/)
- [https://nf-co.re/](https://nf-co.re/)


For your convenience install instructions are reproduced below.


### Nextflow

<a href="https://nextflow.io/"><img src="./images/nextflow2014_no-bg.png" width="300"></a>


All nf-core pipelines use Nextflow, so this must be present on the system where you launch your analysis. See [nextflow.io](https://nextflow.io/) for the latest installation instructions.

Nextflow runs on most POSIX systems (Linux, Mac OSX etc) and can typically be installed by running the following commands:

```
# Make sure that Java v8+ is installed:
java -version

# Install Nextflow
curl -fsSL get.nextflow.io | bash

# Add Nextflow binary to your user's PATH:
mv nextflow ~/bin/
# OR system-wide installation:
# sudo mv nextflow /usr/local/bin
```

You can also install Nextflow using [Bioconda](https://bioconda.github.io/):

```
conda install -c bioconda nextflow
```

### Additional pipeline software

nf-core pipelines utilise Docker, Singularity and Conda to seamlessly install and run different software packages. These packages create isolated compute environments on the fly allowing for flexibility in development and delivery.

 - [Docker](https://docs.docker.com/get-docker/) | 
        Typically used locally / on single-user servers and the cloud.
        Analysis runs in a container, which behaves like an isolated operating system
        Previously required system root access, though a "rootless mode" is available since late 2019
 - [Singularity](https://sylabs.io/docs/) | 
        Often used as an alternative to Docker on multi-user systems such as HPC systems.
        Also runs containers and can create these from Docker images
        Does not need root access or any daemon processes - images built from files
- [Conda](https://docs.conda.io/en/latest/) | 
        Packaging system that manages environments instead of running analysis in containers.
        Poorer reproducibility than Docker / Singularity
            There can be changes in low-level package dependencies over time
            The software still runs in your native operating system environment and so core system functions can differ



### Pipeline code

The nf-core pipeline does not need to be installed. On calling  `nextflow nf-core/rnaseq ...` pipeline components are automatically downloaded from github.


### Reference genomes

Some pipelines come with built-in support for [iGenomes](https://ewels.github.io/AWS-iGenomes) references. It may be preferable for you to download a local copy of these to your system to avoid fetching the same reference many times. For more information, see [Reference genomes](https://nf-co.re/usage/reference_genomes.md).



## Running the nf-core/rnaseq pipeline

Running the nf-core/rnaseq pipeline is incredibly simple and requires only a basic setup.


### Setup


The nf-core/rnaseq pipeline typically requires a sample sheet as input. The sample sheet should comprise the following columns.


|Column| 	Description|
|:--|:--|
|group|Group identifier for sample. This will be identical for replicate samples from the same experimental group.|
|replicate| Integer representing replicate number. Must start from 1..<number of replicates>.|
|fastq_1| 	Full path to FastQ file for read 1. File has to be zipped and have the extension ".fastq.gz" or ".fq.gz".|
|fastq_2| 	Full path to FastQ file for read 2. File has to be zipped and have the extension ".fastq.gz" or ".fq.gz".|
|strandedness| 	Sample strand-specificity. Must be one of unstranded, forward or reverse.|


<br>


A final design file consisting of paired-end data may look something like the one below - see Table 3.1. This samplesheet represents a full sized RNAseq dataset obtained from the ENCODE project, which is routinely used to test the nf-core/rnaseq pipeline and is described [here](https://github.com/nf-core/test-datasets/tree/rnaseq#full-test-dataset-origin)



```{r echo=FALSE}

sampleSheet <- read.delim("RNASeqData/samplesheet.valid.csv", 
                          stringsAsFactors = FALSE, sep = ",")

kableExtra::kable_styling(knitr::kable(
  sampleSheet, booktabs = TRUE, longtable=TRUE, format = "html",
  caption = 'Samplesheet used as input for the nf-core/rnaseq pipeline'
), font_size = 8)


```

## Running the nf-core/rnaseq pipeline

Once the setup is complete running the nf-core/rnaseq pipeline is as simple as typing the following into the command line:



```
  nextflow run nf-core/rnaseq \
      --input samplesheet.csv \
      --fasta '<PATH TO FASTA FILE>/genome.fa'\
      --gtf '<PATH TO FASTA FILE>/genome.gtf'
      --aligner star_rsem
      -profile conda
```



## Quality Control (QC) and first steps


After run completion nf-core will generate a number of data files that can be used to assess sequence quality and for downstream analysis. A typical run folder will comprise the following structure. Please see [Run Folder Example.](https://nf-co.re/rnaseq/results#rnaseq/results-3643a94411b65f42bce5357c5015603099556ad9/star_rsem/)

**QC Metrics**


The first step after the completion of any analysis run is to check the QC metrics. A summary of all  QC metrics for a given run can be found in the file named [multiqc_report.html](RNASeqData/multiqc_report.html) which is located in the MultiQC folder. A detailed overview of the content provided in the MultiQC report can be found at the following location [https://nf-co.re/rnaseq/3.0/output](https://nf-co.re/rnaseq/3.0/output)

<a href="./RNASeqData/multiqc_report.html"><img src="./images/MultiQC.png"></a>


**Merged count files**

After assessing QC we are ready to start with the primary analysis of the data. To do so we will typically use a merged gene count file. For example, [rsem.merged.gene_counts.tsv](RNASeqData/rsem.merged.gene_counts.tsv). These files typically comprise the Ensembl Gene ID, Ensembl Transcript ID and Sample IDs with each column/row entry representing the read count for a given gene. See example below:

<br>

|gene_id|	transcript_id(s)|	GM12878_R1|	GM12878_R2|	H1_R1|	H1_R2|
|:--|:--|:--|:--|:--|:--|:--|:--|
|ENSG00000000003|	ENST00000373020, ...|	7|	0|	8299|	4390|
|ENSG00000000005|	ENST00000373031, ...|	0|	0|	69|	30|		
|ENSG00000000419|	ENST00000371582, ...|	4061|	3582|	1885|	796|
